File: packages/flutter/lib/src/resources/knowledge_sources.dart
/// Knowledge sources resource — `/api/knowledge-sources`.
library;

import 'base.dart';

class KnowledgeSourcesResource extends BaseResource {
  const KnowledgeSourcesResource(super.client);

  Future<List<Map<String, dynamic>>> list() async {
    final data = await client.request('GET', 'knowledge-sources');
    if (data is List) return data.whereType<Map<String, dynamic>>().toList();
    return const [];
  }

  Future<Map<String, dynamic>> get(String id) async {
    final data = await client.request('GET', 'knowledge-sources/$id');
    return (data as Map<String, dynamic>?) ?? {};
  }

  Future<Map<String, dynamic>> create(Map<String, dynamic> payload) async {
    final data =
        await client.request('POST', 'knowledge-sources', body: payload);
    return (data as Map<String, dynamic>?) ?? {};
  }

  Future<Map<String, dynamic>> update(
      String id, Map<String, dynamic> payload) async {
    final data =
        await client.request('PUT', 'knowledge-sources/$id', body: payload);
    return (data as Map<String, dynamic>?) ?? {};
  }

  Future<void> delete(String id) =>
      client.request('DELETE', 'knowledge-sources/$id');

  /// `POST /api/knowledge-sources/{id}/test` — verify connectivity.
  Future<Map<String, dynamic>> test(String id) async {
    final data =
        await client.request('POST', 'knowledge-sources/$id/test');
    return (data as Map<String, dynamic>?) ?? {};
  }

  Future<Map<String, dynamic>> export({String? sourceId}) async {
    final path = sourceId == null
        ? 'knowledge-sources/export'
        : 'knowledge-sources/$sourceId/export';
    final data = await client.request('GET', path);
    return (data as Map<String, dynamic>?) ?? {};
  }

  Future<Map<String, dynamic>> import_(Map<String, dynamic> payload) async {
    final data =
        await client.request('POST', 'knowledge-sources/import', body: payload);
    return (data as Map<String, dynamic>?) ?? {};
  }
}

File: packages/python/src/tbd_agents/resources/knowledge_sources.py
from __future__ import annotations

from typing import Any, Mapping

from tbd_agents.types import (
    ExportedKnowledgeSource,
    ImportResult,
    KnowledgeSource,
    KnowledgeSourceCreate,
    KnowledgeSourceExportBundle,
    KnowledgeSourceImportBundle,
    KnowledgeSourceTestResult,
    KnowledgeSourceUpdate,
)

from .base import BaseResource


class KnowledgeSourcesResource(BaseResource):
    def create(self, payload: KnowledgeSourceCreate | Mapping[str, Any]) -> KnowledgeSource:
        return self._post("/knowledge-sources", payload, model=KnowledgeSource)

    def list(self, *, tags: list[str] | None = None) -> list[KnowledgeSource]:
        params = {"tags": ",".join(tags)} if tags else None
        return self._get_list("/knowledge-sources", params=params, model=KnowledgeSource)

    def get(self, source_id: str) -> KnowledgeSource:
        return self._get(f"/knowledge-sources/{source_id}", model=KnowledgeSource)

    def update(
        self,
        source_id: str,
        payload: KnowledgeSourceUpdate | Mapping[str, Any],
    ) -> KnowledgeSource:
        return self._put(f"/knowledge-sources/{source_id}", payload, model=KnowledgeSource)

    def delete(self, source_id: str) -> None:
        self._delete(f"/knowledge-sources/{source_id}")

    def test(self, source_id: str) -> KnowledgeSourceTestResult:
        return self._post(f"/knowledge-sources/{source_id}/test", model=KnowledgeSourceTestResult)

    def export(self, source_id: str | None = None) -> KnowledgeSourceExportBundle:
        path = (
            "/knowledge-sources/export"
            if source_id is None
            else f"/knowledge-sources/{source_id}/export"
        )
        return self._get(path, model=KnowledgeSourceExportBundle)

    def import_(
        self,
        payload: KnowledgeSourceImportBundle | Mapping[str, Any],
    ) -> ImportResult:
        return self._post("/knowledge-sources/import", payload, model=ImportResult)

